### ISSUE TITLE
Notes from reading

### ISSUE THREAD

**Issue by vitiral** [nikomatsakis/rustc-on-demand-incremental-design-doc#1]
1. The document says says that "queries have no access to mutable state" but that you will "cache the result". Later it shows that the `ctx` object contains the cache (`ctx.query_cache_foo.insert(id, result);`). This language could be improved to say that there is "no mutable state except the cache, which is stored in ctx" -- and then justify the existence of the cache.

2. It seems to me that the `fingerprint` would be cached instead of the `id`, right? Surely the assigment of the `id` is not deterministic? I understand that the `id` can lookup the fingerprint, but this confused me at first.

3. `The HIR(foo) <--- Ty(foo)` example also confused me a bit because I'm wondering what happens with in-lining... isn't the point of incremental compilation that we don't have to recompute LLVM? Are we assuming that `MIR(caller_[1-N])` are never in-lining the function? (If foo was in-lined, wouldn't you have to recompile all dependent nodes?)

**Comment by michaelwoerister** [comment]
> The document says says that "queries have no access to mutable state" but that you will "cache the result". Later it shows that the ctx object contains the cache (ctx.query_cache_foo.insert(id, result);). This language could be improved to say that there is "no mutable state except the cache, which is stored in ctx" -- and then justify the existence of the cache.

I agree. It's important to note that only the query system infrastructure has access to the cache, while someone who defines a new query only specifies the (pure) function that computes the result of a given query and the caching strategy.

> It seems to me that the fingerprint would be cached instead of the id, right? Surely the assigment of the id is not deterministic? I understand that the id can lookup the fingerprint, but this confused me at first.

So, the `id` is indeed deterministic. It is a stable representation of the arguments of a given query invocation. Since a query function must always return the same result for the same arguments, the arguments uniquely identify a given invocation. 

A `fingerprint` on the other hand is a hash of a query's *result*. So if you think of a query as a function it would be like this:
```
fn foo(context: &Context, arg1: T1, arg2: T2, arg3: T3) -> Result
   ^^^                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^     ^^^^^^
The query-kind            The query invocation's ID,       We store a fingerprint
                          unique among all other `foo`     of the result
                          invocations
```
This distinction between `id` and `fingerprint` is what allows us to see if something has changed. Specifically we are determining "has the `foo` query invocation with `id` X a different result `fingerprint` than it had during the last compilation session".

> The HIR(foo) <--- Ty(foo) example also confused me a bit because I'm wondering what happens with in-lining... isn't the point of incremental compilation that we don't have to recompute LLVM? Are we assuming that MIR(caller_[1-N]) are never in-lining the function? (If foo was in-lined, wouldn't you have to recompile all dependent nodes?)

We don't do any inlining at the MIR level yet. When we start doing so, there'll be another query "MIR-after-inlining" and it will have dependency edges to all MIR query nodes that it inlined.

**Comment by eddyb** [comment]
We do have MIR inlining, but it's off by default. I believe the best interaction with the query engine is yet to be decided upon, I've seen some discussion about that but I don't remember where it was.

**Comment by vitiral** [comment]
>  It's important to note that only the query system infrastructure has access to the cache, while someone who defines a new query only specifies the (pure) function that computes the result of a given query and the caching strategy.

ya, that would be a good thing to mention as well.

> So, the id is indeed deterministic.

That is *crazy cool*. I assume it's more than just the argument, for a function it would have to be the crate-path, name and argument types right?

It would be good to provide  quick overfview to what an `id` is and add your explanation to the doc.

> We don't do any inlining at the MIR level yet

Ok, so we can save time on recomputing the **MIR**... but are we saving time on LLVM passes in this stage (is the LLVM output cached as well?)? I'm concerned that we might say "Oh! The type of `foo` hasn't changed so we are fine, no need to re-run LLVM for `caller_3`". The problem is that `caller_3` actually inlined (in the LLVM passes) the old definition of `foo` so it *should have* been recompiled.

I'm not sure how this can be solved without greedily recompiling (in LLVM) the dependency of any node that had it's HIR recomputed.

**Comment by vitiral** [comment]
Also, I might be missing something but I was expecting to see more about the MIR in this doc. Does the MIR have any relevance here?
